### ISSUE TITLE
`needless_match` mistakenly evaluates `match` as no-op

### ISSUE THREAD

**Issue by rdrpenguin04** [rust-lang/rust-clippy#8599]
### Summary

In a standard library implementation I'm helping with, we have a replacement for `Result::as_mut`. We implemented it using `match` ergonomics, and it looks like a no-op because of that. Apparently, Clippy agrees.

### Lint Name

needless_match

### Reproducer

I tried this code:

```rust
pub fn as_mut(&mut self) -> Result<&mut T, &mut E> {
    match self {
        Ok(x) => Ok(x),
        Err(e) => Err(e),
    }
}
```

I saw this happen:

```
error: this match expression is unnecessary
  --> xlang/xlang_abi/src/result.rs:84:9
   |
84 | /         match self {
85 | |             Ok(x) => Ok(x),
86 | |             Err(e) => Err(e),
87 | |         }
   | |_________^ help: replace it with: `self`
```

I expected to see this happen:

Either a valid suggestion or a lack of a lint.

### Version

```text
rustc 1.61.0-nightly (1d9c262ee 2022-03-26)
binary: rustc
commit-hash: 1d9c262eea411ec5230f8a4c9ba50b3647064da4
commit-date: 2022-03-26
host: x86_64-unknown-linux-gnu
release: 1.61.0-nightly
LLVM version: 14.0.0
```

### Additional Labels

@rustbot label +I-suggestion-causes-error
